---
description: SOLID Principles for Clean Architecture
---

# SOLID ì›ì¹™

## ê°œìš”
ëª¨ë“  ì½”ì–´ ë¡œì§ êµ¬í˜„ ì‹œ SOLID ì›ì¹™ì„ ì¤€ìˆ˜í•˜ì—¬ ìœ ì§€ë³´ìˆ˜ì„±, í™•ì¥ì„±, í…ŒìŠ¤íŠ¸ ìš©ì´ì„±ì„ í™•ë³´í•©ë‹ˆë‹¤.

---

## S - Single Responsibility Principle (ë‹¨ì¼ ì±…ì„ ì›ì¹™)

> í´ë˜ìŠ¤ëŠ” ë‹¨ í•˜ë‚˜ì˜ ì±…ì„ë§Œ ê°€ì ¸ì•¼ í•˜ë©°, ë³€ê²½ì˜ ì´ìœ ë„ ë‹¨ í•˜ë‚˜ì—¬ì•¼ í•œë‹¤.

### âœ… Good Example
```javascript
// Calculator.js - ê³„ì‚°ê¸° ìƒíƒœ ê´€ë¦¬ë§Œ ë‹´ë‹¹
class Calculator {
  constructor() {
    this.expression = '';
    this.result = '0';
  }
  
  inputDigit(digit) {
    this.expression += digit;
  }
  
  calculate() {
    const parser = new Parser();
    const evaluator = new Evaluator();
    // íŒŒì‹±ê³¼ í‰ê°€ëŠ” ê°ê°ì˜ í´ë˜ìŠ¤ì— ìœ„ì„
    const tokens = parser.tokenize(this.expression);
    const rpn = parser.parse(tokens);
    this.result = evaluator.evaluate(rpn);
  }
}

// Parser.js - ìˆ˜ì‹ íŒŒì‹±ë§Œ ë‹´ë‹¹
class Parser {
  tokenize(expression) { /* ... */ }
  parse(tokens) { /* ... */ }
}

// Evaluator.js - ìˆ˜ì‹ í‰ê°€ë§Œ ë‹´ë‹¹
class Evaluator {
  evaluate(rpnTokens) { /* ... */ }
}
```

### âŒ Bad Example
```javascript
// Calculator.js - ë„ˆë¬´ ë§ì€ ì±…ì„
class Calculator {
  constructor() {
    this.expression = '';
    this.result = '0';
  }
  
  inputDigit(digit) { /* ... */ }
  
  // íŒŒì‹± ë¡œì§ (ë³„ë„ í´ë˜ìŠ¤ë¡œ ë¶„ë¦¬í•´ì•¼ í•¨)
  tokenize(expression) { /* ... */ }
  parse(tokens) { /* ... */ }
  
  // í‰ê°€ ë¡œì§ (ë³„ë„ í´ë˜ìŠ¤ë¡œ ë¶„ë¦¬í•´ì•¼ í•¨)
  evaluate(rpnTokens) { /* ... */ }
  
  // UI ì—…ë°ì´íŠ¸ (UI ë ˆì´ì–´ì—ì„œ ì²˜ë¦¬í•´ì•¼ í•¨)
  updateDisplay() { /* ... */ }
  
  // ìŠ¤í† ë¦¬ì§€ ì €ì¥ (ë³„ë„ í´ë˜ìŠ¤ë¡œ ë¶„ë¦¬í•´ì•¼ í•¨)
  saveToLocalStorage() { /* ... */ }
}
```

---

## O - Open/Closed Principle (ê°œë°©/íì‡„ ì›ì¹™)

> í™•ì¥ì—ëŠ” ì—´ë ¤ìˆê³ , ìˆ˜ì •ì—ëŠ” ë‹«í˜€ìˆì–´ì•¼ í•œë‹¤.

### âœ… Good Example
```javascript
// MathFunction.js - ê¸°ë³¸ ì¸í„°í˜ì´ìŠ¤
class MathFunction {
  execute(value, angleMode) {
    throw new Error('Must implement execute method');
  }
}

// SinFunction.js
class SinFunction extends MathFunction {
  execute(value, angleMode) {
    const rad = angleMode === 'DEG' ? value * Math.PI / 180 : value;
    return Math.sin(rad);
  }
}

// CosFunction.js
class CosFunction extends MathFunction {
  execute(value, angleMode) {
    const rad = angleMode === 'DEG' ? value * Math.PI / 180 : value;
    return Math.cos(rad);
  }
}

// FunctionRegistry.js - ìƒˆ í•¨ìˆ˜ ì¶”ê°€ ì‹œ ê¸°ì¡´ ì½”ë“œ ìˆ˜ì • ë¶ˆí•„ìš”
class FunctionRegistry {
  constructor() {
    this.functions = new Map();
  }
  
  register(name, functionClass) {
    this.functions.set(name, functionClass);
  }
  
  execute(name, value, angleMode) {
    const func = this.functions.get(name);
    if (!func) throw new Error(`Unknown function: ${name}`);
    return func.execute(value, angleMode);
  }
}

// ì‚¬ìš©
const registry = new FunctionRegistry();
registry.register('sin', new SinFunction());
registry.register('cos', new CosFunction());
// ìƒˆ í•¨ìˆ˜ ì¶”ê°€ ì‹œ ê¸°ì¡´ ì½”ë“œ ìˆ˜ì • ì—†ì´ í™•ì¥ ê°€ëŠ¥
registry.register('tan', new TanFunction());
```

### âŒ Bad Example
```javascript
// Evaluator.js - ìƒˆ í•¨ìˆ˜ ì¶”ê°€ ì‹œ ê¸°ì¡´ ì½”ë“œ ìˆ˜ì • í•„ìš”
class Evaluator {
  applyFunction(func, arg, angleMode) {
    // ìƒˆ í•¨ìˆ˜ ì¶”ê°€í•  ë•Œë§ˆë‹¤ ì´ ë©”ì„œë“œë¥¼ ìˆ˜ì •í•´ì•¼ í•¨
    switch (func) {
      case 'sin':
        const rad1 = angleMode === 'DEG' ? arg * Math.PI / 180 : arg;
        return Math.sin(rad1);
      case 'cos':
        const rad2 = angleMode === 'DEG' ? arg * Math.PI / 180 : arg;
        return Math.cos(rad2);
      // ìƒˆ í•¨ìˆ˜ ì¶”ê°€ ì‹œ ì—¬ê¸°ë¥¼ ìˆ˜ì •í•´ì•¼ í•¨
      case 'tan':
        const rad3 = angleMode === 'DEG' ? arg * Math.PI / 180 : arg;
        return Math.tan(rad3);
      default:
        throw new Error(`Unknown function: ${func}`);
    }
  }
}
```

---

## L - Liskov Substitution Principle (ë¦¬ìŠ¤ì½”í”„ ì¹˜í™˜ ì›ì¹™)

> í•˜ìœ„ íƒ€ì…ì€ ìƒìœ„ íƒ€ì…ì„ ëŒ€ì²´í•  ìˆ˜ ìˆì–´ì•¼ í•œë‹¤.

### âœ… Good Example
```javascript
// Token.js - ê¸°ë³¸ í† í° í´ë˜ìŠ¤
class Token {
  constructor(value) {
    this.value = value;
  }
  
  getValue() {
    return this.value;
  }
}

// NumberToken.js
class NumberToken extends Token {
  constructor(value) {
    super(value);
  }
  
  getValue() {
    return parseFloat(this.value); // ìˆ«ìë¡œ ë³€í™˜
  }
}

// OperatorToken.js
class OperatorToken extends Token {
  constructor(value, precedence) {
    super(value);
    this.precedence = precedence;
  }
  
  getValue() {
    return this.value; // ë¬¸ìì—´ ê·¸ëŒ€ë¡œ ë°˜í™˜
  }
  
  getPrecedence() {
    return this.precedence;
  }
}

// ì‚¬ìš© - Token íƒ€ì…ì„ ê¸°ëŒ€í•˜ëŠ” ê³³ì— í•˜ìœ„ íƒ€ì… ì‚¬ìš© ê°€ëŠ¥
function processToken(token) {
  console.log(token.getValue()); // ëª¨ë“  Token í•˜ìœ„ íƒ€ì…ì—ì„œ ì‘ë™
}

processToken(new Token('test'));
processToken(new NumberToken('42'));
processToken(new OperatorToken('+', 1));
```

### âŒ Bad Example
```javascript
// í•˜ìœ„ í´ë˜ìŠ¤ê°€ ìƒìœ„ í´ë˜ìŠ¤ì˜ ê³„ì•½ì„ ìœ„ë°˜
class Token {
  getValue() {
    return this.value;
  }
}

class BrokenToken extends Token {
  getValue() {
    // ìƒìœ„ í´ë˜ìŠ¤ì™€ ë‹¤ë¥¸ ë™ì‘ (ë°°ì—´ ë°˜í™˜)
    return [this.value, this.type]; // âŒ ê³„ì•½ ìœ„ë°˜
  }
}

// ì‚¬ìš© ì‹œ ë¬¸ì œ ë°œìƒ
function processToken(token) {
  const value = token.getValue();
  console.log(value.toUpperCase()); // BrokenTokenì—ì„œ ì—ëŸ¬ ë°œìƒ
}
```

---

## I - Interface Segregation Principle (ì¸í„°í˜ì´ìŠ¤ ë¶„ë¦¬ ì›ì¹™)

> í´ë¼ì´ì–¸íŠ¸ëŠ” ì‚¬ìš©í•˜ì§€ ì•ŠëŠ” ì¸í„°í˜ì´ìŠ¤ì— ì˜ì¡´í•˜ì§€ ì•Šì•„ì•¼ í•œë‹¤.

### âœ… Good Example
```javascript
// ì‘ì€ ì¸í„°í˜ì´ìŠ¤ë¡œ ë¶„ë¦¬
class Tokenizable {
  tokenize(expression) {
    throw new Error('Must implement tokenize');
  }
}

class Parsable {
  parse(tokens) {
    throw new Error('Must implement parse');
  }
}

// ParserëŠ” ë‘ ì¸í„°í˜ì´ìŠ¤ ëª¨ë‘ êµ¬í˜„
class Parser extends Tokenizable {
  tokenize(expression) { /* ... */ }
}

class ShuntingYardParser extends Parsable {
  parse(tokens) { /* ... */ }
}

// í´ë¼ì´ì–¸íŠ¸ëŠ” í•„ìš”í•œ ì¸í„°í˜ì´ìŠ¤ë§Œ ì˜ì¡´
class ExpressionValidator {
  constructor(tokenizer) {
    this.tokenizer = tokenizer; // Tokenizableë§Œ í•„ìš”
  }
  
  validate(expression) {
    const tokens = this.tokenizer.tokenize(expression);
    // ê²€ì¦ ë¡œì§
  }
}
```

### âŒ Bad Example
```javascript
// ê±°ëŒ€í•œ ì¸í„°í˜ì´ìŠ¤
class MathProcessor {
  tokenize(expression) { /* ... */ }
  parse(tokens) { /* ... */ }
  evaluate(rpn) { /* ... */ }
  format(result) { /* ... */ }
  validate(expression) { /* ... */ }
}

// í´ë¼ì´ì–¸íŠ¸ê°€ í•„ìš” ì—†ëŠ” ë©”ì„œë“œê¹Œì§€ ì˜ì¡´
class ExpressionValidator {
  constructor(processor) {
    this.processor = processor; // MathProcessor ì „ì²´ì— ì˜ì¡´
  }
  
  validate(expression) {
    // tokenizeë§Œ í•„ìš”í•˜ì§€ë§Œ ì „ì²´ ì¸í„°í˜ì´ìŠ¤ì— ì˜ì¡´
    const tokens = this.processor.tokenize(expression);
  }
}
```

---

## D - Dependency Inversion Principle (ì˜ì¡´ì„± ì—­ì „ ì›ì¹™)

> ê³ ìˆ˜ì¤€ ëª¨ë“ˆì€ ì €ìˆ˜ì¤€ ëª¨ë“ˆì— ì˜ì¡´í•˜ì§€ ì•Šì•„ì•¼ í•˜ë©°, ë‘˜ ë‹¤ ì¶”ìƒí™”ì— ì˜ì¡´í•´ì•¼ í•œë‹¤.

### âœ… Good Example
```javascript
// ì¶”ìƒí™” (ì¸í„°í˜ì´ìŠ¤)
class StorageInterface {
  save(key, value) {
    throw new Error('Must implement save');
  }
  
  load(key) {
    throw new Error('Must implement load');
  }
}

// ì €ìˆ˜ì¤€ ëª¨ë“ˆ - LocalStorage êµ¬í˜„
class LocalStorageAdapter extends StorageInterface {
  save(key, value) {
    localStorage.setItem(key, JSON.stringify(value));
  }
  
  load(key) {
    const data = localStorage.getItem(key);
    return data ? JSON.parse(data) : null;
  }
}

// ì €ìˆ˜ì¤€ ëª¨ë“ˆ - SessionStorage êµ¬í˜„
class SessionStorageAdapter extends StorageInterface {
  save(key, value) {
    sessionStorage.setItem(key, JSON.stringify(value));
  }
  
  load(key) {
    const data = sessionStorage.getItem(key);
    return data ? JSON.parse(data) : null;
  }
}

// ê³ ìˆ˜ì¤€ ëª¨ë“ˆ - ì¶”ìƒí™”ì— ì˜ì¡´
class HistoryManager {
  constructor(storage) {
    this.storage = storage; // StorageInterfaceì— ì˜ì¡´
  }
  
  saveHistory(history) {
    this.storage.save('calculator_history', history);
  }
  
  loadHistory() {
    return this.storage.load('calculator_history') || [];
  }
}

// ì‚¬ìš© - êµ¬í˜„ì²´ë¥¼ ì‰½ê²Œ êµì²´ ê°€ëŠ¥
const historyManager1 = new HistoryManager(new LocalStorageAdapter());
const historyManager2 = new HistoryManager(new SessionStorageAdapter());
```

### âŒ Bad Example
```javascript
// ê³ ìˆ˜ì¤€ ëª¨ë“ˆì´ ì €ìˆ˜ì¤€ ëª¨ë“ˆì— ì§ì ‘ ì˜ì¡´
class HistoryManager {
  constructor() {
    // LocalStorageì— ì§ì ‘ ì˜ì¡´ (êµì²´ ë¶ˆê°€ëŠ¥)
  }
  
  saveHistory(history) {
    localStorage.setItem('calculator_history', JSON.stringify(history));
  }
  
  loadHistory() {
    const data = localStorage.getItem('calculator_history');
    return data ? JSON.parse(data) : [];
  }
}

// SessionStorageë¡œ ë³€ê²½í•˜ë ¤ë©´ HistoryManager ì½”ë“œë¥¼ ìˆ˜ì •í•´ì•¼ í•¨
```

---

## ì ìš© ì²´í¬ë¦¬ìŠ¤íŠ¸

### ì½”ë“œ ì‘ì„± ì „
- [ ] ì´ í´ë˜ìŠ¤/í•¨ìˆ˜ì˜ ë‹¨ì¼ ì±…ì„ì´ ëª…í™•í•œê°€?
- [ ] ìƒˆ ê¸°ëŠ¥ ì¶”ê°€ ì‹œ ê¸°ì¡´ ì½”ë“œ ìˆ˜ì •ì´ í•„ìš”í•œê°€?
- [ ] ì˜ì¡´ì„±ì„ ì¶”ìƒí™”ë¡œ í‘œí˜„í•  ìˆ˜ ìˆëŠ”ê°€?

### ì½”ë“œ ë¦¬ë·° ì‹œ
- [ ] ê° í´ë˜ìŠ¤ê°€ í•˜ë‚˜ì˜ ì±…ì„ë§Œ ê°€ì§€ëŠ”ê°€? (SRP)
- [ ] í™•ì¥ ì‹œ ê¸°ì¡´ ì½”ë“œ ìˆ˜ì •ì´ í•„ìš”í•œê°€? (OCP)
- [ ] í•˜ìœ„ íƒ€ì…ì´ ìƒìœ„ íƒ€ì…ì„ ëŒ€ì²´í•  ìˆ˜ ìˆëŠ”ê°€? (LSP)
- [ ] ë¶ˆí•„ìš”í•œ ë©”ì„œë“œì— ì˜ì¡´í•˜ê³  ìˆì§€ ì•Šì€ê°€? (ISP)
- [ ] êµ¬ì²´ì ì¸ êµ¬í˜„ì´ ì•„ë‹Œ ì¶”ìƒí™”ì— ì˜ì¡´í•˜ëŠ”ê°€? (DIP)

### ë¦¬íŒ©í† ë§ ì‹ í˜¸
- ğŸš¨ í´ë˜ìŠ¤ê°€ 100ì¤„ ì´ìƒ â†’ SRP ìœ„ë°˜ ê°€ëŠ¥ì„±
- ğŸš¨ switch/if-elseê°€ ë§ìŒ â†’ OCP ìœ„ë°˜ ê°€ëŠ¥ì„±
- ğŸš¨ í•˜ìœ„ í´ë˜ìŠ¤ì—ì„œ ì˜ˆì™¸ ë°œìƒ â†’ LSP ìœ„ë°˜ ê°€ëŠ¥ì„±
- ğŸš¨ ì‚¬ìš©í•˜ì§€ ì•ŠëŠ” ë©”ì„œë“œê°€ ë§ìŒ â†’ ISP ìœ„ë°˜ ê°€ëŠ¥ì„±
- ğŸš¨ êµ¬ì²´ í´ë˜ìŠ¤ë¥¼ ì§ì ‘ ìƒì„± â†’ DIP ìœ„ë°˜ ê°€ëŠ¥ì„±

---

## ì‹¤ì „ ì˜ˆì‹œ: Calculator í´ë˜ìŠ¤ ë¦¬íŒ©í† ë§

### Before (SOLID ìœ„ë°˜)
```javascript
class Calculator {
  calculate(expression) {
    // íŒŒì‹± (SRP ìœ„ë°˜)
    const tokens = expression.split(' ');
    
    // í‰ê°€ (SRP ìœ„ë°˜)
    let result = 0;
    // ... ë³µì¡í•œ ë¡œì§
    
    // ì €ì¥ (DIP ìœ„ë°˜ - localStorage ì§ì ‘ ì‚¬ìš©)
    localStorage.setItem('last_result', result);
    
    return result;
  }
}
```

### After (SOLID ì¤€ìˆ˜)
```javascript
class Calculator {
  constructor(parser, evaluator, storage) {
    this.parser = parser;       // DIP - ì¶”ìƒí™”ì— ì˜ì¡´
    this.evaluator = evaluator; // DIP - ì¶”ìƒí™”ì— ì˜ì¡´
    this.storage = storage;     // DIP - ì¶”ìƒí™”ì— ì˜ì¡´
  }
  
  calculate(expression) {
    const tokens = this.parser.tokenize(expression);  // SRP - íŒŒì‹± ìœ„ì„
    const rpn = this.parser.parse(tokens);            // SRP - íŒŒì‹± ìœ„ì„
    const result = this.evaluator.evaluate(rpn);      // SRP - í‰ê°€ ìœ„ì„
    this.storage.save('last_result', result);         // SRP - ì €ì¥ ìœ„ì„
    return result;
  }
}
```

---

## ì°¸ê³  ìë£Œ
- [SOLID Principles - Wikipedia](https://en.wikipedia.org/wiki/SOLID)
- [Clean Code by Robert C. Martin](https://www.amazon.com/Clean-Code-Handbook-Software-Craftsmanship/dp/0132350882)
- [Refactoring Guru - SOLID](https://refactoring.guru/design-patterns/solid-principles)
